File: joss/comparisons/cat-2-type-safety/34-enum-validation/example-enum-validation.ts
/**
 * Error Class 34: Enum/Categorical Validation (Runtime)
 *
 * Tidy-TS uses Zod's z.enum() to validate categorical values at data
 * ingestion. If a CSV row contains a value not in the allowed set,
 * the row is rejected immediately with a clear error.
 *
 * Python reads any string value without validation. R's factors
 * accept any level unless explicitly constrained, and unknown levels
 * become NA with a warning.
 */
import { readCSV } from "@tidy-ts/dataframe";
import { z } from "zod";

const Schema = z.object({
  patient_id: z.string(),
  sex: z.enum(["M", "F"]),                        // only M or F
  insurance: z.enum(["Medicare", "Medicaid", "Commercial", "Self-Pay"]),
});

// RUNTIME ERROR if CSV contains sex="X" or insurance="Unknown"
// → Row 5 validation failed: Expected 'M' | 'F', received 'X'

File: joss/comparisons/cat-2-type-safety/34-enum-validation/example-enum-validation.R
# Error Class 34: Enum/Categorical Validation
#
# R's factors with explicit levels produce NA for unknown values —
# with a warning. But read_csv doesn't validate factor levels at all.

library(tidyverse)

# read_csv accepts any value — no factor validation
df <- read_csv(I("patient_id,sex,insurance\nP001,M,Medicare\nP002,X,Unknown\n"),
               show_col_types = FALSE)
print(df)  # X and Unknown silently accepted

# Converting to factor with levels — unknown values become NA (warning)
df$sex <- factor(df$sex, levels = c("M", "F"))
print(df$sex)  # M, NA — X silently became NA
